### PR TITLE
Enforce new grammar rules that `scoped` keyword must be immediately followed by a type possibly “prefixed” by `ref`/`ref readonly`/`in`/`out`.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Related to #62039.
See also https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-07-27.md and https://github.com/dotnet/csharplang/blob/main/proposals/low-level-struct-improvements.md#syntax

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler Please review

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -4691,47 +4700,40 @@ private static bool IsParameterModifier(SyntaxToken token)
...
-        private void ParseParameterModifiers(SyntaxListBuilder modifiers, bool isFunctionPointerParameter = false)
+        private void ParseParameterModifiers(SyntaxListBuilder modifiers, bool isFunctionPointerParameter)
         {
-            while (IsParameterModifier(this.CurrentToken))
+            bool tryScoped = true;
+
+            while (IsParameterModifierExcludingScoped(this.CurrentToken))
             {
-                if (this.CurrentToken.ContextualKind == SyntaxKind.ScopedKeyword)
+                if (this.CurrentToken.Kind is SyntaxKind.RefKeyword or SyntaxKind.OutKeyword or SyntaxKind.InKeyword)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:4715]: >) [](http://example.com/codeflow?start=116&length=1)

`or SyntaxKind.ReadOnlyKeyword` perhaps?

**Review by cston — APPROVED** [review]

**Comment by AlekseyTs** [comment]
https://github.com/orgs/dotnet/teams/roslyn-compiler For the second review

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:4715]: > ```or SyntaxKind.ReadOnlyKeyword``` perhaps?

That is not needed for any valid scenario. And it is really questionable if that would add any value for an error scenario. Going as is for now.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -9827,6 +9845,14 @@ private StatementSyntax ParseLocalDeclarationStatement(SyntaxList<AttributeListS
             var variables = _pool.AllocateSeparated<VariableDeclaratorSyntax>();
             try
             {
+                SyntaxToken scopedKeyword = ParsePossibleScopedKeyword(isFunctionPointerParameter: false);
+
+                // For local functions, 'scoped' is a modifier in LocalFunctionStatementSyntax
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:9850]: Is an error reported if a local function has the scoped modifier?

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:9850]: > Is an error reported if a local function has the scoped modifier?

Yes. There is a test for that. ```ReturnTypeScope``` in RefFieldTest.cs, for example.

**Pull request merged**